### PR TITLE
Java courses Hometask 2

### PR DESCRIPTION
**Pull request by AlexD1991**

### CODE REVIEW

File: src/main/java/controller/StudentsList.java [hunk 1]
```diff
@@ -0,0 +1,35 @@
...
+
+import java.util.ArrayList;
+import java.util.List;
+
+/**
+ * Created by aleksey.dobrovolsky on 6/3/2019.
+ */
+class StudentsList {
+
+    static List<Student> students = init();
```
Comment (dihnatsyeu) [inline src/main/java/controller/StudentsList.java]: even if field is static, good practice is to hide properties ie making them private, and provide getter to access it.
This prevent not relevant changes of the attribute

File: src/main/java/controller/StudentsList.java [hunk 2]
```diff
@@ -0,0 +1,35 @@
...
+
+    private static List<Student> init() {
```
Comment (dihnatsyeu) [inline src/main/java/controller/StudentsList.java]: can you work with students without initializing it? I believe not. This method is the candidate for static block

File: src/main/java/controller/StudentsList.java [hunk 3]
```diff
@@ -0,0 +1,35 @@
...
+        int[] ages = {18, 20, 21, 22, 18, 18, 19, 23, 23, 20};
+        int[] ids = {96, 97, 98, 99, 100, 101, 102, 103, 104, 105};
+
+        for (int i = 0; i < 10; i++) {
+            students.add(new Student(firstNames[i], lastNames[i], ids[i], ages[i]));
+        }
+        return students;
+    }
+
+    static void resultList(List<Student> list) {
```
Comment (dihnatsyeu) [inline src/main/java/controller/StudentsList.java]: this method doesn't belong to the StudentList class. Yes, it prints the list of students but only those that are passed as argument. To make method linked to the StudentList class, it should print the list of students that is property of the class, ie. "students" list. To make it member of a class from the business perspective, print students that is the argument

File: src/main/java/controller/StudentsOperationsClassic.java [hunk 4]
```diff
@@ -0,0 +1,73 @@
...
+import model.Student;
+
+import java.util.*;
+
+/**
+ * Created by aleksey.dobrovolsky on 6/2/2019.
+ */
+public class StudentsOperationsClassic {
+
+    private static List<Student> students = StudentsList.students;
```
Comment (dihnatsyeu) [inline src/main/java/controller/StudentsOperationsClassic.java]: unnecessary reference students to the static object

File: src/main/java/view/Menu.java [hunk 5]
```diff
@@ -0,0 +1,56 @@
...
+import java.io.IOException;
+import java.util.Objects;
+import java.util.Scanner;
+
+/**
+ * Created by aleksey.dobrovolsky on 6/2/2019.
+ */
+public class Menu {
+
+    public static void main(String args[]) throws IOException {
```
Comment (dihnatsyeu) [inline src/main/java/view/Menu.java:16]: StudentOperations hold reference to a static member students, hence once sorted for classics, it is sorted also for streams. Students list shouldn't be static.

File: src/main/java/controller/StudentsOperationsClassic.java [hunk 6]
```diff
@@ -0,0 +1,73 @@
+package controller;
+
+import model.Student;
+
+import java.util.*;
+
+/**
+ * Created by aleksey.dobrovolsky on 6/2/2019.
+ */
+public class StudentsOperationsClassic {
```
Comment (dihnatsyeu) [inline src/main/java/controller/StudentsOperationsClassic.java]: Think how StudentOperations can be organized to get advantages of OOP

File: src/main/java/controller/RateList.java [hunk 7]
```diff
@@ -0,0 +1,98 @@
...
+import java.util.stream.Collectors;
+
+/**
+ * Created by aleksey.dobrovolsky on 6/9/2019.
+ */
+public class RateList {
+
+    private static ArrayList<Rate> rateList = createRateList();
+    private static DecimalFormat df2 = new DecimalFormat("#.##");
+    private static Properties p = new Properties();
```
Comment (dihnatsyeu) [inline src/main/java/controller/RateList.java]: bad name to the variable

File: src/main/java/controller/RateList.java [hunk 8]
```diff
@@ -0,0 +1,98 @@
...
+import java.io.IOException;
+import java.text.DecimalFormat;
+import java.util.ArrayList;
+import java.util.Properties;
+import java.util.stream.Collectors;
+
+/**
+ * Created by aleksey.dobrovolsky on 6/9/2019.
+ */
+public class RateList {
```
Comment (dihnatsyeu) [inline src/main/java/controller/RateList.java]: should all methods of RateList class operate with the class member rateList? If not, methods are not bound to the class

File: src/main/java/controller/RateList.java [hunk 9]
```diff
@@ -0,0 +1,98 @@
...
+
+    private static ArrayList<Rate> rateList = createRateList();
+    private static DecimalFormat df2 = new DecimalFormat("#.##");
```
Comment (dihnatsyeu) [inline src/main/java/controller/RateList.java]: why df2? Anyway, unused property

File: src/main/java/controller/Lists.java [hunk 10]
```diff
@@ -0,0 +1,85 @@
...
+            System.out.println(String.format("ArrayList operation is faster then LinkedList one on %s ms\n", linkedListResult - arrayListResult));
+    }
+
+    private static long getTime(IntConsumer aMethod){
+        long startTime = System.currentTimeMillis();
+        aMethod.accept(0);
+        return System.currentTimeMillis() - startTime;
+    }
+
+
```
Comment (dihnatsyeu) [inline src/main/java/controller/Lists.java]: what is the reason of passing arguments to the below methods if you are using  class static variables in those methods?

File: pom.xml [hunk 11]
```diff
@@ -7,6 +7,18 @@
     <groupId>group</groupId>
     <artifactId>1.0</artifactId>
     <version>1.0-SNAPSHOT</version>
```
Comment (dihnatsyeu) [inline pom.xml:9]: You have used MVC pattern to organize your code. Do you think it suites there? Controller shouldn't hold logic, controllers parse user input and redirect manipulation with objects/logic to model layer.

File: src/main/java/view/Menu.java [hunk 12]
```diff
@@ -0,0 +1,56 @@
...
+                        break;
+                    case "4":
+                        Lists.compare();
+                        break;
+                    case "5":
+                        break;
+                }
+                if (Objects.equals(input, "5")) break;
+
+                System.out.println("Exit? y/n");
```
Comment (dihnatsyeu) [inline src/main/java/view/Menu.java]: why do you need double check to exit?)

File: src/main/java/controller/RateList.java [hunk 13]
```diff
@@ -0,0 +1,98 @@
...
+    private static Properties p = new Properties();
+
+    private static FileReader readProperties() throws FileNotFoundException {
+        return new FileReader(System.getProperty("user.dir") + "/src/main/resources/rateFilterData");
```
Comment (dihnatsyeu) [inline src/main/java/controller/RateList.java]: it's better to read rate properties from classpath - more safe that be bounded to direct location of the file.

File: src/main/java/controller/RateList.java [hunk 14]
```diff
@@ -0,0 +1,98 @@
...
+    }
+
+    private static ArrayList<Rate> sortBySubscriptionFee(ArrayList<Rate> list) {
+        list.sort(Rate.subscriptionFeeComparator);
+        return list;
+    }
+
+    private static ArrayList<Rate> filter(ArrayList<Rate> list) throws IOException {
+        p.load(readProperties());
+        double minMbPerMonth = ((!p.getProperty("minMbPerMonth").isEmpty() ? Double.parseDouble(p.getProperty("minMbPerMonth")) : 0));
```
Comment (dihnatsyeu) [inline src/main/java/controller/RateList.java]: you will get nullpointer exceptions if property is not found in property file

File: src/main/java/controller/RateList.java [hunk 15]
```diff
@@ -0,0 +1,98 @@
...
+        list.add(new Rate4G());
+        list.add(new RateSuper8());
+        list.add(new RateUltra());
+        list.add(new RateInternetMax());
+        list.add(new RateInternetMin());
+        return list;
+
+    }
+
+    public int getClientCount() {
```
Comment (dihnatsyeu) [inline src/main/java/controller/RateList.java]: unused method

File: src/main/java/controller/RateList.java [hunk 16]
```diff
@@ -0,0 +1,98 @@
...
+        controller.RateList.printList(rateList);
+
+        System.out.println("\nSorted by Subscription fee rate list:");
+        controller.RateList.printList(controller.RateList.sortBySubscriptionFee(rateList));
+
+        System.out.println("\nFiltered by the specified condition rate list:");
+        controller.RateList.printList(controller.RateList.filter(rateList));
+    }
+
+    public ArrayList<Rate> getRateList() {
```
Comment (dihnatsyeu) [inline src/main/java/controller/RateList.java]: unused method

File: src/main/java/controller/RateList.java [hunk 17]
```diff
@@ -0,0 +1,98 @@
...
+    private static DecimalFormat df2 = new DecimalFormat("#.##");
+    private static Properties p = new Properties();
+
+    private static FileReader readProperties() throws FileNotFoundException {
+        return new FileReader(System.getProperty("user.dir") + "/src/main/resources/rateFilterData");
+    }
+
+    public static void run() throws IOException {
+        System.out.println("Initial list of rates:");
+        controller.RateList.printList(rateList);
```
Comment (dihnatsyeu) [inline src/main/java/controller/RateList.java]: why don't you just use printList(..) instead of controller.RateList.printList(..). It's overhead.

**Review by dihnatsyeu — CHANGES_REQUESTED** [review]

File: src/main/java/controller/StudentsList.java (hunk 1 above)
Comment (AlexD1991) [reply src/main/java/controller/StudentsList.java]: done - making students private + add getter

File: src/main/java/controller/StudentsList.java (hunk 2 above)
Comment (AlexD1991) [reply src/main/java/controller/StudentsList.java]: done - remove redundant students attribute

File: src/main/java/controller/RateList.java (hunk 7 above)
Comment (AlexD1991) [reply src/main/java/controller/RateList.java]: done - remove redundant attribute df2

File: src/main/java/controller/RateList.java (hunk 9 above)
Comment (AlexD1991) [reply src/main/java/controller/RateList.java]: done - remove redundant attribute df2

File: src/main/java/controller/StudentsList.java (hunk 1 above)
Comment (AlexD1991) [reply src/main/java/controller/StudentsList.java]: done - changed the logic of StudentsList class. Removed static content from the class, all methods change attributes related to the objects of the class. 
students getter became redundant

File: src/main/java/controller/StudentsList.java (hunk 2 above)
Comment (AlexD1991) [reply src/main/java/controller/StudentsList.java]: done - init is called in constructor

File: src/main/java/controller/StudentsList.java (hunk 3 above)
Comment (AlexD1991) [reply src/main/java/controller/StudentsList.java]: done - print is not static right now and works with the list of attributes related to the object

File: src/main/java/controller/StudentsOperationsClassic.java (hunk 4 above)
Comment (AlexD1991) [reply src/main/java/controller/StudentsOperationsClassic.java]: done - removed logic of static methods. The class was removed as redundant

File: src/main/java/view/Menu.java (hunk 5 above)
Comment (AlexD1991) [reply src/main/java/view/Menu.java:16]: done - removed logic of static methods. The class was removed as redundant

File: src/main/java/controller/StudentsOperationsClassic.java (hunk 6 above)
Comment (AlexD1991) [reply src/main/java/controller/StudentsOperationsClassic.java]: done - removed logic of static methods. The class was removed as redundant

File: src/main/java/controller/RateList.java (hunk 8 above)
Comment (AlexD1991) [reply src/main/java/controller/RateList.java]: removed not related methods to new classes - RateListRunner and RateListFilterProperties

File: pom.xml (hunk 11 above)
Comment (AlexD1991) [reply pom.xml:9]: done - removed controller package at all 
All methods are incuded to the related classes and located in model package
view is only for menu displaying

File: src/main/java/controller/RateList.java (hunk 13 above)
Comment (AlexD1991) [reply src/main/java/controller/RateList.java]: done

File: src/main/java/controller/RateList.java (hunk 14 above)
Comment (AlexD1991) [reply src/main/java/controller/RateList.java]: done - made some changes to handle of the file missed situation and properties missed cases

File: src/main/java/controller/RateList.java (hunk 15 above)
Comment (AlexD1991) [reply src/main/java/controller/RateList.java]: done - the method was not called previously. Now it's called in RateListRunner

File: src/main/java/controller/RateList.java (hunk 16 above)
Comment (AlexD1991) [reply src/main/java/controller/RateList.java]: done - removed

File: src/main/java/controller/RateList.java (hunk 17 above)
Comment (AlexD1991) [reply src/main/java/controller/RateList.java]: done - printList is called as a method of StudetsList objects without any attributes

File: src/main/java/view/Menu.java (hunk 12 above)
Comment (AlexD1991) [reply src/main/java/view/Menu.java]: done- removed an extra cycle

File: src/main/java/controller/Lists.java (hunk 10 above)
Comment (AlexD1991) [reply src/main/java/controller/Lists.java]: was not changed
It's for using methods as arguments of this method 
So the method gets another method as attribute, executes it and calculates the execution time

**Pull request merged**